rs/nns/governance/src/governance/disburse_maturity.rs

```diff
@@ -137,6 +139,65 @@ impl TryFrom<Account> for Icrc1Account {
     }
 }
 
+impl Destination {
+    fn try_new(
+        account: &Option<Account>,
+        account_identifier: &Option<AccountIdentifierProto>,
+        caller: PrincipalId,
+    ) -> Result<Self, String> {
+        let destination = match (account, account_identifier) {
+            (Some(account), None) => Destination::AccountToDisburseTo(account.clone()),
+            (None, Some(account_identifier_proto)) => {
+                Destination::AccountIdentifierToDisburseTo(account_identifier_proto.clone())
+            }
+            (None, None) => Destination::AccountToDisburseTo(Account {
+                owner: Some(caller),
+                subaccount: None,
+            }),
+            (Some(_), Some(_)) => {
+                return Err("Cannot provide both to_account and to_account_identifier".to_string())
+            }
+        };
+        // We make sure we only construct a destination that can be converted to a valid account identifier.
+        let _ = destination.account_identifier()?;
+        Ok(destination)
+    }
+
+    fn account_identifier(&self) -> Result<AccountIdentifier, String> {
+        match self {
+            Destination::AccountToDisburseTo(account) => {
+                let icrc1_account = Icrc1Account::try_from(account.clone())?;
+                Ok(AccountIdentifier::from(icrc1_account))
+            }
+            Destination::AccountIdentifierToDisburseTo(account_identifier_proto) => {
+                AccountIdentifier::try_from(account_identifier_proto)
+                    .map_err(|_| "Invalid account identifier".to_string())
+            }
+        }
+    }
+
+    /// Returns the account to disburse to, if it is specified as `to_account`.
+    pub fn to_account(&self) -> Option<Account> {
+        match self {
+            Destination::AccountToDisburseTo(account) => Some(account.clone()),
+            Destination::AccountIdentifierToDisburseTo(_) => None,
+        }
+    }
+
+    /// Returns the account identifier to disburse to, if it is specified as
+    /// `to_account_identifier`. Note that even though `to_account` can be converted to an account
+    /// identifier, this method does not perform that conversion but simply returns what the user
+    /// initially specified.
+    pub fn to_account_identifier(&self) -> Option<AccountIdentifierProto> {
```

Comment:
This is actually kind of surprising, since Account -> AccountIdentifier is always valid, but AccountIdentifier -> Account is not.

It's odd to have "account_identifier" and "to_account_identifier" because they're very similarly named and you might expect them to do the same thing...  but if they were going to be different, I'd expect them to do the opposite of what they actually do.  

account_identifier() sounds more like an accessor and to_account_identifier sounds more like it would do conversion.

wdyt?